torch_geometric/transforms/rooted_subgraphs.py

```diff
@@ -0,0 +1,209 @@
+            return -1
+        else:
+            return super().__cat_dim__(key, value, *args, **kwargs)
+
+
+class RootedSubgraph(BaseTransform):
+    r"""
+    Base class for rooted subgraph.
+    The object transforms a Data object to RootedSubgraphData object.
+    """
+    def extract_subgraph(self, data: Data):
+        r""" For a input graph with N nodes, extract a rooted subgraph for every node in the graph.
+        Return:
+            subgraph_nodes_mask: N x N dense mask matrix, i-th row indicates the rooted subgraph of
+                node i.
+            hop_indicator: N x N dense matrix, i-th row indicates the distance from other nodes to
+                node i.
+        """
+        raise NotImplementedError
+
+    def __call__(self, data: Data) -> Data:
+        subgraph_nodes_mask, hop_indicator_dense = self.extract_subgraphs(
+            data)
+        subgraph_edges_mask = subgraph_nodes_mask[:, data.edge_index[0]] & \
+                               subgraph_nodes_mask[:, data.edge_index[1]]  # N x E dense mask matrix
+        subgraph_nodes, subgraph_edges, hop_indicator = to_sparse(
+            subgraph_nodes_mask, subgraph_edges_mask, hop_indicator_dense)
+
+        combined_subgraph = combine_subgraph(data.edge_index,
+                                             subgraph_nodes,
+                                             subgraph_edges,
+                                             num_selected=data.num_nodes,
+                                             num_nodes=data.num_nodes)
+
+        data = RootedSubgraphData(**{k: v for k, v in data})
+        data.subgraph_batch = subgraph_nodes[0]
+        data.subgraph_nodes_mapper = subgraph_nodes[1]
+        data.subgraph_edges_mapper = subgraph_edges[1]
+        data.combined_subgraph = combined_subgraph
+        data.hop_indicator = hop_indicator
+        return data
+
+
+class RootedEgoNets(RootedSubgraph):
+    """ Record rooted k-hop Egonet for each node in the graph.
+    From the `"From Stars to Subgraphs: Uplifting Any GNN with Local Structure Awareness"
+    <https://arxiv.org/pdf/2110.03753.pdf>`_ paper
+
+    Args:
+        num_hops (int): k for k-hop Egonet.
+    """
+    def __init__(self, num_hops: int):
+        super().__init__()
+        self.num_hops = num_hops
+
+    def extract_subgraphs(self, data: Data):
+        # return k-hop subgraph for all nodes in the graph
+        row, col = data.edge_index
+        sparse_adj = SparseTensor(
+            row=row, col=col, sparse_sizes=(data.num_nodes, data.num_nodes))
+        hop_mask = sparse_adj.to_dense() > 0
+        hop_indicator = torch.eye(data.num_nodes, dtype=torch.int,
+                                  device=data.edge_index.device) - 1
+        for i in range(self.num_hops):
+            hop_indicator[(hop_indicator == -1) & hop_mask] = i + 1
+            hop_mask = sparse_adj.matmul(hop_mask.float()) > 0
+
+        hop_indicator = hop_indicator.T  # N x N
```

Comment:
I think we should follow neighborhood sampling/message passing convention here and sample source nodes. I think this implementation can also be made more more efficient by just using `torch.ops.torch_sparse.neighbor_sample` directly.

Reply:
Okay aligned.

As for using `torch.ops.torch_sparse.neighbor_sample` - I'm not clear on how this would work, unless we did a `torch.ops.torch_sparse.neighbor_sample` for each node in the graph separately? 

It would also lose the distance (`hop_indicator`) information, though it seems optional in the RW graph at least.

Reply:
Good point. Let's leave it out for now :)

Reply:
It might still be worth converting the existing logic to sparse matrix multiplication? Thoughts?

Reply:
Yes, we can add it as a TODO.